### ISSUE TITLE
RD3 : NullPointerException on session timeout with LDAP

### ISSUE THREAD

**Issue by Daryes** [rundeck/rundeck#3939]
**Describe the bug**
This error occurs in the log, but it has no noticeable effect on Rundeck itself, I'm able to reconnect with the same user without trouble.

```
2018-08-28 17:24:07.698  WARN --- [tp1659209382-34] .s.a.j.DefaultJaasAuthenticationProvider : Error error logging out of LoginContext

javax.security.auth.login.LoginException: java.lang.NullPointerException
        at org.eclipse.jetty.jaas.spi.AbstractLoginModule.logout(AbstractLoginModule.java:298)
```

**My Rundeck detail**

* Rundeck version: 3.0.4
* install type: deb
* OS Name/version: ubuntu 16.04
* DB Type/version: H2 file

**To Reproduce**
1. Use a mixed local auth + Active Directory auth (conf at the end)
2. login successfully with an AD user.
3. wait for the session timeout (currently 60 min)
4. check the log for the Null Exception error, with an erroneous timestamp : 
the reported time is login time +2m but won't appear in the log before the timeout occurs.

Full log after a Rundeck service start (sensitive data edited between #<># )

```
2018-08-28 17:21:13.987  INFO --- [           main] rundeckapp.BootStrap                     : Rundeck startup finished in 967ms
Grails application running at http://localhost:4440 in environment: production
2018-08-28 17:22:09.615 DEBUG --- [tp1659209382-56] ailsUsernamePasswordAuthenticationFilter : Request is to process authentication
2018-08-28 17:22:09.743  INFO --- [tp1659209382-56] c.d.r.j.j.JettyCachingLdapLoginModule    : Login attempts: 1, Hits: 0, Ratio: 0%.
2018-08-28 17:22:09.744 DEBUG --- [tp1659209382-56] c.d.r.j.j.JettyCachingLdapLoginModule    : Cache Miss for #<AD USER>#.
2018-08-28 17:22:09.744 DEBUG --- [tp1659209382-56] c.d.r.j.j.JettyCachingLdapLoginModule    : Searching for users with filter: '(&(objectClass={0})({1}={2}))' from base dn: #<DOMAIN BASE DN>#
2018-08-28 17:22:09.760 DEBUG --- [tp1659209382-56] c.d.r.j.j.JettyCachingLdapLoginModule    : Found user?: true
2018-08-28 17:22:09.762  INFO --- [tp1659209382-56] c.d.r.j.j.JettyCachingLdapLoginModule    : Attempting authentication: #<USER CN>#
2018-08-28 17:22:09.774 DEBUG --- [tp1659209382-56] c.d.r.j.j.JettyCachingLdapLoginModule    : JettyCachingLdapLoginModule: User '#<AD USER>#' has roles: [user]
2018-08-28 17:22:09.780 DEBUG --- [tp1659209382-56] c.d.r.j.j.JettyCachingLdapLoginModule    : Adding #<AD USER># set to expire: 1535469729780300000
2018-08-28 17:22:09.810 DEBUG --- [tp1659209382-56] ailsUsernamePasswordAuthenticationFilter : Authentication success. Updating SecurityContextHolder to contain: org.springframework.security.authentication.jaas.JaasAuthenticationToken@c9c33391: Principal: #<USER LOGIN>#; Credentials: [PROTECTED]; Authenticated: true; Details: org.springframework.security.web.authentication.WebAuthenticationDetails@ffff4c9c: RemoteIpAddress: 192.168.0.222; SessionId: node01xg7lh9h42mac93gyrc497l290; Granted Authorities: Jaas Authority [user,user]
2018-08-28 17:24:07.698  WARN --- [tp1659209382-34] .s.a.j.DefaultJaasAuthenticationProvider : Error error logging out of LoginContext

javax.security.auth.login.LoginException: java.lang.NullPointerException
        at org.eclipse.jetty.jaas.spi.AbstractLoginModule.logout(AbstractLoginModule.java:298)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at javax.security.auth.login.LoginContext.invoke(LoginContext.java:755)
        at javax.security.auth.login.LoginContext.access$000(LoginContext.java:195)
        at javax.security.auth.login.LoginContext$4.run(LoginContext.java:682)
        at javax.security.auth.login.LoginContext$4.run(LoginContext.java:680)
        at java.security.AccessController.doPrivileged(Native Method)
        at javax.security.auth.login.LoginContext.invokePriv(LoginContext.java:680)
        at javax.security.auth.login.LoginContext.logout(LoginContext.java:628)
        at org.springframework.security.authentication.jaas.AbstractJaasAuthenticationProvider.handleLogout(AbstractJaasAuthenticationProvider.java:266)
        at org.springframework.security.authentication.jaas.AbstractJaasAuthenticationProvider.onApplicationEvent(AbstractJaasAuthenticationProvider.java:282)
        at org.springframework.security.authentication.jaas.AbstractJaasAuthenticationProvider.onApplicationEvent(AbstractJaasAuthenticationProvider.java:116)
        at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:172)
        at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:165)
        at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:139)
        at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:393)
        at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:347)
        at org.springframework.security.web.session.HttpSessionEventPublisher.sessionDestroyed(HttpSessionEventPublisher.java:91)
        at org.eclipse.jetty.server.session.SessionHandler.removeSession(SessionHandler.java:1061)
        at org.eclipse.jetty.server.session.SessionHandler.invalidate(SessionHandler.java:1234)
        at org.eclipse.jetty.server.session.DefaultSessionIdManager.invalidateAll(DefaultSessionIdManager.java:486)
        at org.eclipse.jetty.server.session.Session.invalidate(Session.java:854)
        at org.springframework.security.web.authentication.logout.SecurityContextLogoutHandler.logout(SecurityContextLogoutHandler.java:66)
        at grails.plugin.springsecurity.web.authentication.logout.MutableLogoutFilter$_doFilter_closure1.doCall(MutableLogoutFilter.groovy:71)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:98)
        at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
        at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:264)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1034)
        at groovy.lang.Closure.call(Closure.java:418)
        at groovy.lang.Closure.call(Closure.java:434)
        at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2125)
        at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2110)
        at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2151)
        at grails.plugin.springsecurity.web.authentication.logout.MutableLogoutFilter.doFilter(MutableLogoutFilter.groovy:71)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at grails.plugin.springsecurity.web.SecurityRequestHolderFilter.doFilter(SecurityRequestHolderFilter.groovy:58)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214)
        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:177)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
        at org.grails.web.servlet.mvc.GrailsWebRequestFilter.doFilterInternal(GrailsWebRequestFilter.java:77)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
        at org.grails.web.filters.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:67)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
        at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:197)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
        at org.springframework.web.filter.CorsFilter.doFilterInternal(CorsFilter.java:96)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
        at org.springframework.boot.actuate.autoconfigure.MetricsFilter.doFilterInternal(MetricsFilter.java:106)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:533)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:146)
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:257)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1595)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1317)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1564)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1219)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
        at org.eclipse.jetty.server.Server.handle(Server.java:531)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:352)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:260)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:281)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
        at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
        at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:762)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:680)
        at java.lang.Thread.run(Thread.java:748)
```

Aside the error, reconnecting with the same user doesn't raise any issue in the GUI.

**JASS configuration**
Loaded through env variables : 
```
JAAS_CONF="/etc/rundeck/jaas-combinedldap.conf"
LOGIN_MODULE="multiauth"
```
jaas-combinedldap.conf
```
multiauth {
// first login module : usual rundeck mode
        org.eclipse.jetty.jaas.spi.PropertyFileLoginModule sufficient
        debug="true"
        file="/etc/rundeck/realm.properties"
        // required for combined login
        storePass="true";


// second login module : AD
        // com.dtolabs.rundeck.jetty.jaas.JettyCachingLdapLoginModule required
        com.dtolabs.rundeck.jetty.jaas.JettyCombinedLdapLoginModule required
        debug="true"
        contextFactory="com.sun.jndi.ldap.LdapCtxFactory"

        providerUrl="ldap://#<AD DOMAIN>#:389"
        bindDn="#<AD DOMAIN BIND USER>#"
        bindPassword="#<BIND USER PASSWORD>#"

        authenticationMethod="simple"
        forceBindingLogin="true"

        userBaseDn="#<AD USER BASE DN>#"
        userRdnAttribute="sAMAccountName"
        userIdAttribute="sAMAccountName"
        userPasswordAttribute="unicodePwd"

        userObjectClass="user"

        roleBaseDn="#<AD GROUP STARTING DN>#"
        roleNameAttribute="cn"
        roleMemberAttribute="member"
        roleObjectClass="group"

        cacheDurationMillis="300000"
        reportStatistics="true"

// custom AD args
        nestedGroups="true"
        rolePrefix="rundeck_"
        clearPass="true"
        // connection timeout (ms)
        timeoutRead="10000"
        timeoutConnect="20000"
        // default role for authenticated users
        supplementalRoles="user";
};
```

- edited to remove a missed domain login

**Comment by gschueler** [comment]
I think I've seen this before, happening upon logout.  as you say, it doesn't cause any issue, but does produce ugly message in the log. I believe it is an error in Jetty code, but I haven't fully investigated it

**Comment by ProTip** [comment]
Fixed by #4647
